homeassistant/components/sia/alarm_control_panel.py

```diff
@@ -0,0 +1,253 @@
+    "CB": STATE_ALARM_ARMED_AWAY,
+    "CG": STATE_ALARM_ARMED_AWAY,
+    "CL": STATE_ALARM_ARMED_AWAY,
+    "CP": STATE_ALARM_ARMED_AWAY,
+    "CQ": STATE_ALARM_ARMED_AWAY,
+    "CS": STATE_ALARM_ARMED_AWAY,
+    "CF": STATE_ALARM_ARMED_CUSTOM_BYPASS,
+    "OA": STATE_ALARM_DISARMED,
+    "OB": STATE_ALARM_DISARMED,
+    "OG": STATE_ALARM_DISARMED,
+    "OP": STATE_ALARM_DISARMED,
+    "OQ": STATE_ALARM_DISARMED,
+    "OR": STATE_ALARM_DISARMED,
+    "OS": STATE_ALARM_DISARMED,
+    "NC": STATE_ALARM_ARMED_NIGHT,
+    "NL": STATE_ALARM_ARMED_NIGHT,
+    "BR": PREVIOUS_STATE,
+    "NP": PREVIOUS_STATE,
+    "NO": PREVIOUS_STATE,
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: Callable[..., None],
+) -> bool:
+    """Set up SIA alarm_control_panel(s) from a config entry."""
+    async_add_entities(
+        [
+            SIAAlarmControlPanel(entry, account_data, zone)
+            for account_data in entry.data[CONF_ACCOUNTS]
+            for zone in range(
+                1,
+                entry.options[CONF_ACCOUNTS][account_data[CONF_ACCOUNT]][CONF_ZONES]
+                + 1,
+            )
+        ]
+    )
+    return True
+
+
+class SIAAlarmControlPanel(AlarmControlPanelEntity, RestoreEntity):
+    """Class for SIA Alarm Control Panels."""
+
+    def __init__(
+        self,
+        entry: ConfigEntry,
+        account_data: dict[str, Any],
+        zone: int,
+    ):
+        """Create SIAAlarmControlPanel object."""
+        self._entry: ConfigEntry = entry
+        self._account_data: dict[str, Any] = account_data
+        self._zone: int = zone
+
+        self._port: int = self._entry.data[CONF_PORT]
+        self._account: str = self._account_data[CONF_ACCOUNT]
+        self._ping_interval: int = self._account_data[CONF_PING_INTERVAL]
+
+        self.entity_id: str = ALARM_ENTITY_ID_FORMAT.format(
+            SIA_ENTITY_ID_FORMAT.format(
+                self._port, self._account, self._zone, DEVICE_CLASS_ALARM
+            )
+        )
+
+        self._attr: dict[str, Any] = {
+            CONF_PORT: self._port,
+            CONF_ACCOUNT: self._account,
+            CONF_ZONE: self._zone,
+            CONF_PING_INTERVAL: f"{self._ping_interval} minute(s)",
+        }
+
+        self._available: bool = True
+        self._state: StateType = None
+        self._old_state: StateType = None
+        self._cancel_availability_cb: CALLBACK_TYPE | None = None
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass.
+
+        Overridden from Entity.
+
+        1. start the event listener and add the callback to on_remove
+        2. get previous state from storage
+        3. if previous state: restore
+        4. if previous state is unavailable: set _available to False and return
+        5. if available: create availability cb
+        """
+        self.async_on_remove(
+            self.hass.bus.async_listen(
```

Comment:
Yeah, I had that discussion with @bdraco, the reason I am using the event bus for this is that there are several dozen different codes defined in the SIA protocol, only a small subset is used by the alarm control panel platforms, several others will be used by binary_sensors and sensors (in subsequent PR's for this integration). But different alarm system manufacturers use the other codes that aren't covered for other actions, and a user might want to take a different actions based on the specific code, where this platform takes the codes and maps them to a HA state, for both of those other purposes I want the user to be able to make their own automations (or trigger-based entities) so that they can have the full benefit without it having to all be covered in this code.

Reply:
We can still fire a user facing event. But the internal communication should be handled by our dispatch helper.

Reply:
ok, I have a new branch created, will setup a draft PR in a couple of minutes, will do this.